File: ui-api/src/main/java/io/tipsters/uiapi/web/MatchesController.java
package io.tipsters.uiapi.web;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

import javax.validation.Valid;

import io.tipsters.uiapi.dto.CompetitionMatches;
import io.tipsters.uiapi.dto.MatchesRequest;
import io.tipsters.uiapi.service.MatchesService;

@RestController
@RequestMapping("/matches")
public class MatchesController {

  private final MatchesService matchesService;

  public MatchesController(MatchesService matchesService) {
    this.matchesService = matchesService;
  }

  @PostMapping
  public List<CompetitionMatches> matchesByCompetition(@Valid @RequestBody MatchesRequest matchesRequest) {
    return matchesService.matchesByCompetitions(matchesRequest);
  }
}

File: web-server/src/main/kotlin/io/tipsters/controller/MatchesController.kt
package io.tipsters.controller

import io.tipsters.common.data.MatchesByCompetition
import io.tipsters.service.MatchService
import org.hibernate.validator.constraints.NotEmpty
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod.POST
import org.springframework.web.bind.annotation.RestController
import java.time.LocalDateTime
import java.util.*
import javax.validation.Valid

@RestController
@RequestMapping("/matches")
class MatchesController @Autowired constructor(val matchService: MatchService) {

    @RequestMapping(method = arrayOf(POST))
    fun matchesByCompetitions(@Valid @RequestBody request: MatchesByCompetitions): List<MatchesByCompetition> {
        return matchService.matchesByCompetitions(request)
    }
}

data class MatchesByCompetitions(
    @get:NotEmpty
    val competitionIDs: Set<UUID>,

    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    val start: LocalDateTime = LocalDateTime.MIN,

    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    val end: LocalDateTime = LocalDateTime.MAX
)
